src/actions/tab.ts

```diff
@@ -0,0 +1,84 @@
+import {Action} from './index';
+import {PlainReduxAction, ReduxAction} from './redux-action';
+import {RESULT_ACTION_TYPE_INDEX, ResultAction} from './result';
+import {ShelfAction, SPEC_ACTION_TYPE_INDEX} from './shelf';
+
+/* Actions concerning a single tab */
+export type SingleTabAction = (
+  ResultAction |
+  ShelfAction |
+  TitleUpdate
+);
+
+export type SingleTabActionType = SingleTabAction['type'];
+
+export const SINGLE_TAB_ACTION_TYPE_INDEX: {[k in SingleTabActionType]: 1} = {
+  ...RESULT_ACTION_TYPE_INDEX,
+
+  FILTER_ADD: 1,
+  FILTER_CLEAR: 1,
+  FILTER_MODIFY_EXTENT: 1,
+  FILTER_MODIFY_MAX_BOUND: 1,
+  FILTER_MODIFY_MIN_BOUND: 1,
+  FILTER_MODIFY_TIME_UNIT: 1,
+  FILTER_MODIFY_ONE_OF: 1,
+  FILTER_REMOVE: 1,
+  FILTER_TOGGLE: 1,
+
+  SHELF_AUTO_ADD_COUNT_CHANGE: 1,
+  SHELF_GROUP_BY_CHANGE: 1,
+  SHELF_LOAD_QUERY: 1,
+
+  ...SPEC_ACTION_TYPE_INDEX,
+
+  TITLE_UPDATE: 1
+};
+
+export function isSingleTabAction(action: Action): action is SingleTabAction {
+  return SINGLE_TAB_ACTION_TYPE_INDEX[action.type];
+}
+
+export const TITLE_UPDATE = 'TITLE_UPDATE';
+export type TitleUpdate = ReduxAction<typeof TITLE_UPDATE, {
+  newTitle: string;
+}>;
+
+/* Actions concerning multiple tabs */
+export type MultiTabAction = (
+  TabAdd |
+  TabRemove |
+  TabSwitch
+);
+
+export type MultiTabActionType = MultiTabAction['type'];
+
+export const MULTI_TAB_ACTION_TYPE_INDEX: {[k in MultiTabActionType]: 1} = {
+  TAB_ADD: 1,
+  TAB_REMOVE: 1,
+  TAB_SWITCH: 1
+};
+
+export function isMultiTabAction(action: Action): action is MultiTabAction {
+  return MULTI_TAB_ACTION_TYPE_INDEX[action.type];
+}
+
+export const TAB_ADD = 'TAB_ADD';
+export type TabAdd = PlainReduxAction<typeof TAB_ADD>;
+
+export const TAB_REMOVE = 'TAB_REMOVE';
+export type TabRemove = PlainReduxAction<typeof TAB_REMOVE>;
+
+export const TAB_SWITCH = 'TAB_SWITCH';
+export type TabSwitch = ReduxAction<typeof TAB_SWITCH, {
+  tabID: number;
+}>;
+
+/* Union of tab actions */
+export type TabAction = SingleTabAction | MultiTabAction;
```

Comment:
Actually maybe we should have a folder structure  

`src/actions/plot-screen/index.ts`, `xxx.ts` (move actions for one plot-screen inside there)

Reply:
In my app, a tab can be any of the three "genres": 

tab = {chart, dashboard, story}.

In Voyager, how about we say:

tab = {view, dashboard, story}, or tab = {plot, dashboard, story}

The Voyager interface has "specified view", "related views" and `ViewPane`, so I think it will make sense to call the default tab genre `views` or `viewTabs`. On the other hand, voyager also has `Plot` and `PlotList`, so I think we can also call the default tab genre `plots` or `plotTabs`.

Even though Voyager doesn't have dashboards and stories (yet), I'm trying to imagine the state and folder structure in the context of three genres. You can decide if this makes sense for Voyager:

The state could be like this:

```js
persistent: ...
undoable:
  present:
    views:
      activeTabID: number
      viewTabs: [ViewTabState] // i.e., rename SingleViewTabState to ViewTabState
    dashboards:
      activeTabID: number
      dashboardTabs: [DashboardTabState]
    stories:
      activeTabID: number
      storyTabs: [StoryTabState]
```

The actions folder could be like this:

```
src/actions/views
  index.ts 
  view-tabs.ts // contains VIEW_TAB_ADD, VIEW_TAB_REMOVE, VIEW_TAB_SWITCH
  result.ts
  shelf.ts
  title.ts // contains TITLE_UPDATE
src/actions/dashboards
  index.ts 
  dashboard-tabs.ts // contains DASHBOARD_TAB_ADD, DASHBOARD_TAB_REMOVE, DASHBOARD_TAB_SWITCH
  ... // some within-dashboard actions
src/actions/stories
  index.ts 
  story-tabs.ts // contains STORY_TAB_ADD, STORY_TAB_REMOVE, STORY_TAB_SWITCH
  ... // some within-story actions
```

The reducers folder could be like this:

```
src/reducers/views
  view-tabs.ts // handles VIEW_TAB_ADD, VIEW_TAB_REMOVE etc.
  view-tabs.test.ts
  result.ts
  result.test.ts
  shelf.ts
  shelf.test.ts
  title.ts
  title..test.ts
src/reducers/dashboards
  dashboard-tabs.ts // handles DASHBOARD_TAB_ADD etc.
  ...
src/reducers/stories
  story-tabs.ts // handles STORY_TAB_ADD etc.
  ...
```

Reply:
The term view is quite confusing.  Plot is btter.  

For our tab model, we only need one array like we already have, not multiple arrays.  Please don't change this.   We already discussed this before... You can have your own set of extra arrays as your organize tabs by type of tabs, not single list of array like in traditional tabs UI.

> On the other hand, voyager also has Plot and PlotList

That's why I want to add "screen" suffix.   Your dashboard will have similar problem too.
You will have a screen for authoring dashboard, which has the actual dashboard component 
surrounded by other component for editing dashboard. So that's why having suffix like screen probably will solve the same problem similarly.